Fix Hash(c10::Scalar), account for garbage data in union

wconstab opened the pull request:
Summary:
Hash(c10::Scalar) made a bad assumption that it was valid to just hash over all the bytes of data of the c10::Scalar struct.

Becuase c10::Scalar stores a union of different (float/int/complex) types with different sizes, not all bytes are valid in all cases.  Hash() should only read the bytes corresponding to the currently active type.

Test Plan: Added new unit tests.  Verified HashTest.Scalar failed with the original Hash() impl and then fixed.

Differential Revision: D32367564

alanwaketan commented on torch/csrc/lazy/core/hash.h:
```diff
@@ -74,7 +74,18 @@ static inline hash_t Hash(const c10::ScalarType& value) {
+  case c10::ScalarType::Bool:
+    return Hash(value.toBool());
+  default:
+    throw std::runtime_error("Unknown scalar type.");
```
Maybe better to use TORCH_CHECK if we want users to know the error or TORCH_INTERNAL_ASSERT.

alanwaketan commented on test/cpp/lazy/test_misc.cpp:
```diff
@@ -11,9 +11,29 @@ template <typename T>
+  // simulate some garbage in the unused bits of the
+  // the tagged union that is c10::Scalar, which is bigger
+  // than the size of the int64_t we're currently using it with
+  *((uint8_t*)&b)  = 1;
```
Can we do `static_cast<uint8_t*>` here? Can you explain the memory layout here a bit such that it's easier for me to understand why writing to "offset 0" makes sense? Is there any way we can tell the differences from `a` to `b` such that we can also verify that this operation succeeds?

wconstab replied:
There isn't any way to verify it without violating the API of the Scalar.  I didn't know the memory layout either, except by experimenting.  I found it's sizeof() was 32 bytes, which surprised me.  I don't know exactly what memory i'm stepping on here, but, for the purposes of this experiment it's not important as long as it isn't memory used by the `Long` member field, which it isn't.

I did verify it though, by observing that the below EXPECTs all failed when I clobber this memory without fixing the hash function.

I don't see the point of changing the cast.  I don't benefit from any compiler safety checks here, in fact, i'm doing something 'unsafe' on purpose.

wconstab replied:
Just for you, I decided to try this :) 
But it doesn't appear to work:
`Static_cast from 'c10::Scalar *' to 'uint8_t *' (aka 'unsigned char *') is not allowed`

Anyway, regular cast is fine.

alanwaketan replied:
Yea, my bad. It should be reinterpret_cast for your use case:
`*(reinterpret_cast<uint8_t*>(&b))  = 1;`.

alanwaketan replied:
Did a little research. It seems non-trivial to find any resources to answer me the memory layout question. So, it's fine to keep the comment intact.

wconstab committed:
Fix Hash(c10::Scalar), account for garbage data in union (#68201)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/68201

Hash(c10::Scalar) made a bad assumption that it was valid to just hash over all the bytes of data of the c10::Scalar struct.

Becuase c10::Scalar stores a union of different (float/int/complex) types with different sizes, not all bytes are valid in all cases.  Hash() should only read the bytes corresponding to the currently active type.

Test Plan: Added new unit tests.  Verified HashTest.Scalar failed with the original Hash() impl and then fixed.

Differential Revision: D32367564

fbshipit-source-id: be1dbc4932890ebb70898184483b6776b068c6b0
